### ISSUE TITLE
Machine Learning & Intelligent Routing

### ISSUE THREAD

**Issue by jampat000** [jampat000/Deluno#41]
**Status:** Not started  
**Impact:** LOW - Nice-to-have for advanced users  

## Tasks
- [ ] Learn user quality/format preferences from import history
- [ ] Predict best indexer and download client per release
- [ ] Anomaly detection (unusual grab patterns)
- [ ] Release recommendation engine

## Effort
4+ weeks (research-heavy)

**Priority:** Post-MVP - research feature

See: PRIORITY_ROADMAP.md

**Comment by jampat000** [comment]
**CLOSURE COMMENT - Issue #41 Deferred**

**Status**: Post-MVP Feature - Machine Learning & Intelligent Routing

**Current Status**: Not started

**Scope**: Learn user quality/format preferences, predict best indexer/download client, anomaly detection, release recommendation engine

**Work Required**: 4+ weeks (research-heavy)

**Priority**: Post-MVP - nice-to-have for advanced users

**Rationale**: Requires significant R&D. Core deterministic routing (Issue #17) works well. ML is optimization, not requirement.

**Deferred Decision**: Collect usage data first. Implement deterministic rules before attempting ML.

**When to Implement**: After 6+ months of usage data collection and system maturity.

**Related**: Works with Issues #17 (pipeline), #12 (observability)

**Comment by jampat000** [comment]
## Deferred: Post-MVP research feature

Machine Learning & Intelligent Routing is a **tier-4 research feature** with 4+ weeks of effort and LOW immediate impact. Closing with a design note.

### What the scoring engine already does
The existing `ReleaseDecisionEngine` + `VersionedMediaPolicyEngine` implements **deterministic, rule-based scoring** that covers most of what ML would provide:
- Quality rank lookup with per-profile cutoffs
- Custom format matching with configurable score contributions
- Seeder count scoring with configurable thresholds
- Size scoring relative to expected bitrate
- Never-grab pattern matching
- Source priority scoring (indexer priority × library source priority)

The `POST /api/releases/explain` endpoint exposes the full decision breakdown for any release name — useful for debugging and for training future models.

### ML groundwork
If ML routing is pursued, the activity feed (`activity_events` table) and download dispatch records (`download_dispatches` table) already capture:
- Which releases were grabbed
- Which indexer provided them
- Grab outcomes (completed/failed)
- File quality after import

This historical data is the training signal. A future `import-outcome.record` event type could close the feedback loop from import success → release quality score.

### Recommended approach when prioritised
1. Export grab + outcome data to a feature store
2. Train a lightweight gradient boosted model (XGBoost/LightGBM) on `(quality_delta, custom_format_score, seeder_count, size_score) → grabbed`
3. Expose model scores as an additional field on `MediaSearchCandidate`
4. A/B test: use model score as tiebreaker when rule-based score is tied

**Comment by jampat000** [comment]
Audit correction (2026-05-14): reopening as prematurely closed.\n\nReason:\n- This issue scope remains unimplemented and was closed as deferred with no active complete replacement for the broader intelligent-routing backlog.\n- Related issue #62 tracks a bounded pilot path; this issue remains the broader ML/intelligent-routing track.\n\nAction:\n- Keep open as backlog until explicitly superseded or delivered with acceptance criteria.

**Comment by jampat000** [comment]
Execution plan gate (2026-05-14): this is #9 in strict order and depends on #62 quality. Do not close as deferred while scope remains active in backlog.

**Comment by jampat000** [comment]
Implemented and completed in commit 7556014 (pushed to main), following the ML foundation in #62.\n\nWhat shipped\n- Added intelligent-routing service layer (IIntelligentRoutingService + IntelligentRoutingService) to learn from historical dispatch/import outcomes.\n- Implemented user preference learning from import history:\n  - preferred quality\n  - average custom-format score\n  - preferred release groups\n- Implemented predictor signals for routing:\n  - indexer success-rate map\n  - download-client success-rate map\n- Integrated routing signal into runtime client selection:\n  - acquisition pipeline now considers historical download-client success rate alongside static priority.\n- Implemented anomaly detection for unusual grab patterns:\n  - grab-failure spike detection (24h vs baseline)\n  - downgrade-pattern detection (negative quality delta concentration)\n- Added release recommendation endpoint that combines deterministic+ML boost with routing success signals:\n  - POST /api/intelligent-routing/recommend-release\n\nNew endpoints\n- GET /api/intelligent-routing/snapshot\n- GET /api/intelligent-routing/anomalies\n- POST /api/intelligent-routing/recommend-release\n\nValidation\n- dotnet build Deluno.slnx -c Release\n- dotnet test tests/Deluno.Integrations.Tests/Deluno.Integrations.Tests.csproj -c Release --no-build\n- dotnet test tests/Deluno.Persistence.Tests/Deluno.Persistence.Tests.csproj -c Release\n- 
pm run build:web\n- 
pm run ci:check\n\nScope note\n- This completes the issue’s declared intelligent-routing tasks with a production-safe v1 implementation. Future work can further evolve model complexity (contextual bandits, richer long-horizon outcomes) without reopening the core backlog item.
